aggregator/src/aggregator.rs

```diff
@@ -1932,345 +1940,329 @@ impl VdafOps {
-                                // which is a whole-aggregation problem rather than a per-report
-                                // problem. (separately, this would require more than u64::MAX
-                                // report shares in a single aggregation job, which is practically
-                                // impossible.)
-                                ord.try_into()?,
-                                Some(PrepareResp::new(
+                                ),
+                            };
+
+                        sender.send(ReportShareData {
+                            report_share: prepare_init.report_share().clone(),
+                            report_aggregation: WritableReportAggregation::new(
+                                ReportAggregation::<SEED_SIZE, A>::new(
+                                    *task.id(),
+                                    aggregation_job_id,
                                     *prepare_init.report_share().metadata().id(),
-                                    prepare_step_result,
-                                )),
-                                report_aggregation_state,
+                                    *prepare_init.report_share().metadata().time(),
+                                    // Unwrap safety: we checked that all ordinal values are representable
+                                    // as a u64 before entering the parallel iterator.
+                                    ord.try_into().unwrap(),
+                                    Some(PrepareResp::new(
+                                        *prepare_init.report_share().metadata().id(),
+                                        prepare_step_result,
+                                    )),
+                                    report_aggregation_state,
+                                ),
+                                output_share,
                             ),
-                            output_share,
-                        ),
-                    });
-                }
-                Ok(report_share_data)
+                        })
+                    })
+            }
+        });
+
+        let mut report_share_data = Vec::with_capacity(req.prepare_inits().len());
+        while receiver.recv_many(&mut report_share_data, 10).await > 0 {}
+
+        // Await the producer task to resume any panics that may have occurred, and to ensure we can
+        // unwrap the aggregation parameter's Arc in a few lines. The only other errors that can
+        // occur are: a `JoinError` indicating cancellation, which is impossible because we do not
+        // cancel the task; and a `SendError`, which can only happen if this future is cancelled (in
+        // which case we will not run this code at all).
+        let _ = producer_task.await.map_err(|join_error| {
+            if let Ok(reason) = join_error.try_into_panic() {
+                panic::resume_unwind(reason);
             }
-        };
```

Comment:
I wonder if we should add a defensive check that report_share_data is the right length, if the blocking task didn't panic.

Reply:
Sure. I used an `assert_eq!` as it would be programmer error if this condition did not hold.